exporters/otlp/otlptrace/otlptracegrpc/client.go

```diff
@@ -17,92 +17,259 @@ package otlptracegrpc // import "go.opentelemetry.io/otel/exporters/otlp/otlptra
+//
+// Any active connections to a remote endpoint are closed if they were created
+// by the client. Any gRPC connection passed during creation using
+// WithGRPCConn will not be closed. It is the caller's responsibility to
+// handle cleanup of that resource.
+//
+// This method synchronizes with the UploadTraces method of the client. It
+// will wait for any active calls to that method to complete unimpeded, or it
+// will cancel any active calls if ctx expires. If ctx expires, the context
+// error will be forwarded as the returned error. All client held resources
+// will still be released still in this situation.
+//
+// If the client has already stopped, an error will be returned describing
+// this.
 func (c *client) Stop(ctx context.Context) error {
-	return c.connection.Shutdown(ctx)
+	// Acquire the c.tscMu lock within the ctx lifetime.
+	acquired := make(chan struct{})
+	go func() {
+		c.tscMu.Lock()
+		close(acquired)
+	}()
+	var err error
+	select {
+	case <-ctx.Done():
+		// The Stop timeout is reached. Kill any remaining exports to force
+		// the clear of the lock and save the timeout error to return and
+		// signal the shutdown timed out before cleanly stopping.
+		c.stopFunc()
+		err = ctx.Err()
+
+		// To ensure the client is not left in a dirty state c.tsc needs to be
+		// set to nil. To avoid the race condition when doing this, ensure
+		// that all the exports are killed (initiated by c.stopFunc).
+		<-acquired
+	case <-acquired:
+	}
+	// Hold the tscMu lock for the rest of the function to ensure no new
+	// exports are started.
+	defer c.tscMu.Unlock()
+
+	// The otlptrace.Client interface states this method is called only
+	// once, but there is no guarantee it is called after Start. Ensure the
+	// client is started before doing anything and let the called know if they
+	// made a mistake.
+	if c.tsc == nil {
+		return errAlreadyStopped
+	}
+
+	// Clear c.tsc to signal the client is stopped.
+	c.tsc = nil
+
+	if c.ourConn {
+		closeErr := c.conn.Close()
+		// A context timeout error takes precedence over this error.
+		if err == nil && closeErr != nil {
+			err = closeErr
+		}
+	}
+	return err
 }
 
-// UploadTraces sends a batch of spans to the collector.
+var errShutdown = errors.New("the client is shutdown")
+
+// UploadTraces sends a batch of spans.
+//
+// Retryable errors from the server will be handled according to any
+// RetryConfig the client was created with.
 func (c *client) UploadTraces(ctx context.Context, protoSpans []*tracepb.ResourceSpans) error {
-	if !c.connection.Connected() {
-		return fmt.Errorf("traces exporter is disconnected from the server %s: %w", c.connection.SCfg.Endpoint, c.connection.LastConnectError())
+	// Hold a read lock to ensure a shut down initiated after this starts does
+	// not abandon the export. This read lock acquire has less priority than a
+	// write lock acquire (i.e. Stop), meaning if the client is shutting down
+	// this will come after the shut down.
+	c.tscMu.RLock()
```

Comment:
Just to double-check. Do we expect to have a lot of concurrent `UploadTraces`? I am not sure why we are using a `sync.RWMutex` instead of `sync.Mutex`. _I never analyzed the Go's reader-writer mutex implementation, but e.g. in .NET it is a LOT heavier than a simple mutex._

Reply:
I was using a `RWLock` here as the read lock algorithm (allowing multiple concurrently, and prioritizing a write lock) matched the behavior we want here. There is indeed a performance hit for high concurrency scaling on many core systems (https://github.com/golang/go/issues/17973). I'm not entirely sure this will affect the common use case. If we are worried about this performance though, I can take a look at possibly using an atomic pointer here instead.

Reply:
If multiple parallel reads are desired then leave it as it is and resolve the comment. Performance can be improved in a separate PR if needed. BTW impressive work